File: app/src/main/java/remix/myplayer/ui/fragment/base/BaseMusicFragment.java
package remix.myplayer.ui.fragment.base;

import android.content.Context;
import android.os.Bundle;
import android.view.View;

import remix.myplayer.helper.MusicEventHelper;
import remix.myplayer.ui.activity.base.BaseMusicActivity;

public class BaseMusicFragment extends BaseFragment implements MusicEventHelper.MusicEventCallback{
    private BaseMusicActivity mMusicActivity;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mMusicActivity = (BaseMusicActivity) context;
        } catch (ClassCastException e) {
            throw new RuntimeException(context.getClass().getSimpleName() + " must be an instance of " + BaseMusicActivity.class.getSimpleName());
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mMusicActivity = null;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        MusicEventHelper.addCallback(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        MusicEventHelper.removeCallback(this);
    }

    @Override
    public void onMediaStoreChanged() {

    }

    @Override
    public void onPermissionChanged(boolean has) {

    }

    @Override
    public void onPlayListChanged() {

    }

    @Override
    public void onServiceConnected() {
    }

    @Override
    public void onServiceDisConnected() {
    }
}

File: app/src/main/java/remix/myplayer/ui/fragment/base/BaseMusicFragment.kt
package remix.myplayer.ui.fragment.base

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.viewbinding.ViewBinding
import remix.myplayer.bean.mp3.Song
import remix.myplayer.helper.MusicEventCallback

import remix.myplayer.service.MusicService
import remix.myplayer.ui.activity.base.BaseMusicActivity

abstract class BaseMusicFragment<VB: ViewBinding> : BaseFragment(), MusicEventCallback {
  protected abstract val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> VB
  private var _binding: VB? = null
  protected val binding
    get() = _binding!!

  private var musicActivity: BaseMusicActivity? = null

  override fun onAttach(context: Context) {
    super.onAttach(context)
    try {
      musicActivity = context as BaseMusicActivity?
    } catch (e: ClassCastException) {
      throw RuntimeException(context.javaClass.simpleName + " must be an instance of " + BaseMusicActivity::class.java.simpleName)
    }

  }

  override fun onDetach() {
    super.onDetach()
    musicActivity = null
  }

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View? {
    _binding = bindingInflater.invoke(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
//        MusicEventHelper.addCallback(this)
    musicActivity?.addMusicServiceEventListener(this)
  }

  override fun onDestroyView() {
    super.onDestroyView()
//        MusicEventHelper.removeCallback(this)
    musicActivity?.removeMusicServiceEventListener(this)
    _binding = null
  }

  override fun onMediaStoreChanged() {

  }

  override fun onPermissionChanged(has: Boolean) {
    hasPermission = has
  }

  override fun onPlayListChanged(name: String) {

  }

  override fun onMetaChanged() {
  }

  override fun onPlayStateChange() {
  }

  override fun onTagChanged(oldSong: Song, newSong: Song) {

  }

  override fun onServiceConnected(service: MusicService) {}

  override fun onServiceDisConnected() {}
}
